common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java

```diff
@@ -0,0 +1,222 @@
+                                if (mPositiveButtonListener != null) {
+                                    mPositiveButtonListener.onClick(certDetails);
+                                }
+                            }
+                        })
+                        //Negative button should end up cancelling flow.
+                        .setNegativeButton(R.string.smartcard_cert_dialog_negative_button, new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                //On request by user, cancel flow.
+                                if (mNegativeButtonListener != null) {
+                                    mNegativeButtonListener.onClick();
+                                }
+                            }
+                        });
+                // Create dialog.
+                final AlertDialog alertDialog = builder.create();
+                // Set up single checked item logic for cert ListView within dialog.
+                final ListView listView = alertDialog.getListView();
+                listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+                    @Override
+                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                        listView.setItemChecked(position,true);
+                        certAdapter.notifyDataSetChanged();
+                    }
+                });
+
+                //If user touches outside dialog, the default behavior makes the dialog disappear without really doing anything.
+                //Adding this line in disables this default behavior so that the user can only exit by hitting the cancel button.
+                alertDialog.setCanceledOnTouchOutside(false);
+                //Handle back button the same as the negative button.
+                alertDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    @Override
+                    public void onCancel(DialogInterface dialog) {
+                        if (mNegativeButtonListener != null) {
+                            mNegativeButtonListener.onClick();
+                        }
+                    }
+                });
+                mDialog = alertDialog;
+            }
+        });
+    }
+
+    /**
+     * Handles scenario when CBA is canceled unexpectedly (for example. when a YubiKey is unplugged while a dialog is showing).
+     */
+    @Override
+    public void onCancelCba() {
+        //Call CancelCbaCallback's onCancel
+        mCancelCbaCallback.onCancel();
+        //Dismiss dialog
+        dismiss();
+    }
+
+    /**
+     * Sets listener for positive button.
+     * @param listener Implemented PositiveButtonListener.
+     */
+    public void setPositiveButtonListener(PositiveButtonListener listener) {
```

Comment:
These listeners are needed to be re-assigned? and hence providing setters for these? Can these instead be supplied on object construction for this class?

Reply:
The same question for all the listeners

Reply:
I think these can be supplied on object construction. I'll make these changes for all the listeners